src/shell.js

```diff
@@ -21,6 +21,12 @@ var Module = typeof {{{ EXPORT_NAME }}} !== 'undefined' ? {{{ EXPORT_NAME }}} :
 #endif // USE_CLOSURE_COMPILER
 #endif // SIDE_MODULE
 
+// Environment defaults for use by --pre-jses. See environment setup code below.
```

Comment:
comment seems to suggest they are only for use by --pre-jses. How about removing this comment, and in the comment later down, changing it to `--pre-jses are emitted after the Module integration code, so that they can refer to Module (if they choose; they can also define Module). They are also emitted after the ENVIRONMENT_* detection code so that they can use those variables.`

Reply:
Well, it's not *only*, but *especially*, because pre-jses do not yet see the implications of `Module['ENVIRONMENT']` if present, and have the option to modify `Module['ENVIRONMENT']` theirself (as they always had), resulting in different values than the defaults further down the road.

Not sure how common this is, but the sheer presence of `Module['ENVIRONMENT']` makes this ugly, because moving up more than just the defaults *possibly* breaks existing programs that set or modify `Module['ENVIRONMENT']` in pre-jses even though they didn't know the defaults yet, while moving up just the defaults introduces a lot of edge cases.

I agree that the comment should be more clear, though.

Thinking about it a little more: With the presence of environment detection in pre-jses now, I'd even go as far as to remove `Module['ENVIRONMENT']` entirely if it's not absolutely necessary for something (I don't know the full picture), leaving such things to the pre-jses which can freely modify whatever they want, using whatever custom settings. To me, `Module['ENVIRONMENT']` seems to be redundant with all this being possible in pre-jses, and removing it would make things  more intuitive and consistent. Convention over configuration, basically.

Reply:
That's a good point.

Historically I think we added that option because some special environments like PhantomJS were not detected properly. But we fixed a lot of that, so maybe it isn't even needed now. Still, odd environments will exist...

I like the idea of deprecating `Module['ENVIRONMENT']` with this PR, since it allows both reading and writing to the `ENVIRONMENT_*` vars from pre-jses. Let's see if anyone has any concerns with that, but if not, we can deprecate it with

 * Clarifying in the comment that using a pre-js is the proper way to modify the environment detection.
 * When ASSERTIONS are on, check if `Module['ENVIRONMENT']` is set, and if so issue an error about it not checked anymore and a pre-js being the proper way.